[Tests] Include branch coverage info in coverage test

achow101 opened the pull request:
This PR  adds an option to configure to include branch coverage statistics be gathered during `make cov`. It also disables logging when coverage is enabled so that the logging branches are not included in the coverage stats. This has a side effect of causing `-Wunused-variable` warnings during compile time when coverage is enabled.

achow101 committed:
Have `make cov` optionally include branch coverage statistics

Added an option to configure to allow for branch coverage statistics gathering.

Disabled logprint macro when coverage testing is on so that unnecessary branches are not analyzed.

sipa commented:
Here is a possible hack to force a variable list of arguments to be treated as 'used' for the purpose of `-Wunused-variable`:

```c++
static inline void MarkUsed() {}
template<typename T, typename... Args> static inline void MarkUsed(const T& t, const Args&... args)
{
    (void)t;
    MarkUsed(args...);
}

#ifdef USE_COVERAGE
#define LogPrintf(...) do { MarkUsed(__VA_ARGS__); } while(0)
#define LogPrint(category, ...) do { MarkUsed(__VA_ARGS__); } while(0)
#else
```

achow101 commented:
Used @sipa's hack to shut up the warnings.

gmaxwell commented:
Cool. Concept ACK.

achow101 committed:
Replace lcov -r commands with faster way

Instead of using lcov -r (which is extremely slow), first use a python script to perform bulk cleanup of the /usr/include/* coverage. Then use lcov -a to remove the duplicate entries. This has the same effect of lcov -r but runs significantly faster

sipa commented on contrib/filter-lcov.py:
```diff
@@ -0,0 +1,24 @@
+with open(tracefile, 'r') as f:
+    with open(outfile, 'w') as wf:
+        for line in f:
+            if pattern in line:
```
Perhaps instead use `line.startswith("SF:" + pattern)` instead? That would prevent matching on variables that accidentally match the pattern name.

achow101 replied:
Done

sipa commented on contrib/filter-lcov.py:
```diff
@@ -0,0 +1,24 @@
+                in_remove = True
+            if not in_remove:
+                wf.write(line)
+            if 'end_of_record' in line:
```
Perhaps use `line == 'end_of_record'` instead? Same reason as above.

achow101 replied:
done

sipa commented:
utACK 405b86a92aee4f2ddb6710bfe07ff714f2afcfa2

MarcoFalke commented on contrib/filter-lcov.py:
```diff
@@ -0,0 +1,24 @@
+#!/usr/bin/env python3
```
explicit license is missing:
```py
# Copyright (c) 2017 Bitcoin Core Developers
# Distributed under the MIT software license, see the accompanying
# file COPYING or http://www.opensource.org/licenses/mit-license.php.

achow101 replied:
done

MarcoFalke commented on contrib/filter-lcov.py:
```diff
@@ -0,0 +1,24 @@
+    with open(outfile, 'w') as wf:
+        for line in f:
+            if line.startswith("SF:") and pattern in line:
+                in_remove = True
```
Not sure why you need the var `in_remove`. Couldn't you write `wf.write(line);continue;` here?

achow101 replied:
The format of the file is

```
SF:<path to file>
<bunch of lines>
end_of_record
<repeat>
```
In order to know that we are in a `<bunch of lines>` section that should be removed (because the `SF:...` line matches our removal pattern, we need to have a variable to remember to remember that we shouldn't be writing these lines to the output file. That is what `in_remove` is for.

maflcko reviewed (commented):
Concept ACK

theuni commented:
The lcov stuff was added _ages_ ago, when lcov's --no-external was too new to rely on. I think it'd be ok now to require >=1.10 and forego all of the filtering.

MarcoFalke commented:
@achow101 Are you still working on this?

achow101 commented:
@MarcoFalke Yes

achow101 commented:
Would it be okay with everyone if I just merged #10565 into this PR? It all falls under "improvements to coverage data".

sipa commented:
@achow101 Have you tried the approach @theuni suggested above (using --no-external, and removing all filtering)?

achow101 commented:
@sipa I am thinking of doing that and combining #10565 into this PR. If I remove filtering here, it would just be added back in by #10565.

MarcoFalke commented:
Needs rebase.

achow101 commented:
Closing this as #10565 contains this PR (I totally forgot) except for the copyright header I added last commit :/
